### ISSUE TITLE
Callback on_fit_epoch_end called twice at the end of model.train

### ISSUE THREAD

**Issue by moritzmeister** [ultralytics/ultralytics#18735]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

Hi!

I noticed that the `on_fit_epoch_end` callback event gets triggered also at the end of a training run in the final_eval: https://github.com/ultralytics/ultralytics/blob/d9292fb7f10a2ee0265af28fad5d56b8ab79eece/ultralytics/engine/trainer.py#L689

I am wondering if that's intended. This leads to problems when using the call back for example log mlflow metrics, because the eval metric is being logged in the same graph as the metrics during training, with a duplicate epoch:
![Image](https://github.com/user-attachments/assets/c449502e-c7cf-4e5c-8947-452bc9933f5b)

Note I am not sure how this looks like if you're using the ultralytics mlflow integration, we decided not to use it because we failed to log everything in one run, ultralytics would always create a seperate run even though we run training within an mlflow run context already. So I think there might be another problem.

Could this be changed to have a separate callback event for this?
Happy to contribute if you agree.

### Environment

```
Ultralytics 8.3.62 🚀 Python-3.12.3 torch-2.4.0+cu124 CUDA:0 (Tesla T4, 14931MiB)
Setup complete ✅ (4 CPUs, 10.8 GB RAM, 36.8/146.6 GB disk)

OS                  Linux-5.15.0-1075-aws-x86_64-with-glibc2.39
Environment         Linux
Python              3.12.3
Install             pip
RAM                 10.83 GB
Disk                36.8/146.6 GB
CPU                 Intel Xeon Platinum 8259CL 2.50GHz
CPU count           4
GPU                 Tesla T4, 14931MiB
GPU count           1
CUDA                12.4

numpy               ✅ 1.26.4>=1.23.0
numpy               ✅ 1.26.4<2.0.0; sys_platform == "darwin"
matplotlib          ✅ 3.8.4>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 10.3.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.32.2>=2.23.0
scipy               ✅ 1.13.1>=1.4.1
torch               ✅ 2.4.0+cu124>=1.8.0
torch               ✅ 2.4.0+cu124!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.19.0+cu124>=0.9.0
tqdm                ✅ 4.66.4>=4.64.0
psutil              ✅ 5.9.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 1.5.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.14>=2.0.0
```

### Minimal Reproducible Example
Note you have to have a local mlflow tracking server running.

```
import mlflow
from ultralytics import YOLO, settings

def main():
    remote_server_uri = "http://localhost:8080"  # set to your server URI
    mlflow.set_tracking_uri(remote_server_uri)
    mlflow.set_experiment("/my-experiment")

    mlflow.pytorch.autolog(disable=False)
    settings.update({"mlflow": False})

    def sanitize_dict(x):
        """Sanitize dictionary keys by removing parentheses and converting values to floats."""
        return {"cb_" + k.replace("(", "").replace(")", ""): float(v) for k, v in x.items()}
    
    def on_fit_epoch_end(trainer):
        print(f"{trainer.epoch}, {trainer.metrics}")
        if mlflow:
            mlflow.log_metrics(metrics=sanitize_dict(trainer.metrics), step=trainer.epoch)

    with mlflow.start_run():
        # Create a new YOLO model from scratch
        model = YOLO("yolo11n.pt")

        # Load a pretrained YOLO model (recommended for training)
        model.add_callback("on_fit_epoch_end", on_fit_epoch_end)

        # Train the model using the 'coco8.yaml' dataset for 3 epochs
        results = model.train(data="coco8.yaml", epochs=3)

if __name__ == "__main__":
    main()

```

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @moritzmeister, thank you for bringing this up and for your interest in contributing to the Ultralytics project 🚀!

We recommend checking out the [Docs](https://docs.ultralytics.com) for valuable information about training, integrations, and implementation details. If you're new to our framework, resources like [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples may prove helpful.

Regarding your 🐛 Bug Report, to ensure we can investigate effectively, could you please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) that demonstrates the issue? It seems you've described the situation well, but a concise code snippet can help us debug the callback behavior more efficiently.

If your report involves MLflow integration issues, logging inconsistencies, or specific use cases, providing any configurations, related code examples, or environment details (which you’ve included—thank you! 🙌) will also be very helpful.

### Upgrade
Ensure you're using the latest `ultralytics` release and dependencies. Upgrade in a [**Python>=3.8**](https://www.python.org/) environment:

```bash
pip install -U ultralytics
```

This ensures compatibility with the latest updates and bug fixes. You’re currently using version 8.3.62, but it's always good to double-check for updates.

### Join the Community
To discuss your use case and gather insights or feedback from the Ultralytics community, feel free to join us:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time conversations
- [Discourse](https://community.ultralytics.com) for in-depth discussions
- [Subreddit](https://reddit.com/r/Ultralytics) for broader sharing and exchanges

### Verified Environments
YOLO can run in various verified environments with all dependencies preinstalled. For quick GPU access, see:
- **Notebooks**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Cloud and Servers**: [GCP](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/), [AWS](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/), or [Docker](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker"></a>

## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing.

Lastly, note that this is an automated response 🤖, but rest assured an Ultralytics engineer will review and assist with your concern soon. We appreciate your willingness to contribute—let us know if you have specific pull request ideas to address the callback behavior! 🚀

**Comment by moritzmeister** [comment]
Hi @UltralyticsAssistant, and everyone else!

I added a minimal reproducible example now. See above.

**Comment by Y-T-G** [comment]
Can you open a PR with the fix?

**Comment by moritzmeister** [comment]
Sure, I can @Y-T-G, but do you agree it should be a separate callback event? Or rather remove the call to the `on_fit_epoch_end` callback at the end of the final eval?

**Comment by Y-T-G** [comment]
You can remove it from final_eval
